docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,114 @@
+
+## Proposal
+
+In order to support key-value stores, the following need to be added to Spin:
+
+- A `WIT` file that defines the key-value interface
+- SDK implementations for various programming languages
+- A default local key-value store
+
+Although not in scope for this proposal, we'll also want to expand the runtime configuration code added in [this PR](https://github.com/fermyon/spin/pull/798) to support configuring various key-value stores, including the default one.
+
+### Key-Value Interface (`.wit`)
+
+Spin should leverage the WebAssembly WASI subgroup’s work to define a universal key-value `WIT` interface. That work is taking place in [the `wasi-keyvalue` proposal](https://github.com/WebAssembly/wasi-keyvalue). However, the proposal was made assuming that `WIT` star imports are available, which is not yet the case. Therefore, we're modeling stores as pseudo-[resource handles](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource) which may be created using an `open` function and disposed using a `close` function.  Each operation on the store is a function which accepts a handle as its first parameter.
+
+```
+// A handle to an open key-value store
+type store = u32
+
+// The set of errors which may be raised by functions in this interface
+variant error {
+  // Too many stores have been opened simultaneously. Closing one or more
+  // stores prior to retrying may address this.
+  store-table-full,
+
+  // The host does not recognize the store name requested.  Defining and
+  // configuring a store with that name in a runtime configuration file
+  // may address this.
+  no-such-store,
+
+  // The store handle provided is not recognized, i.e. it was either never
+  // opened or has been closed.
+  invalid-store,
+
+  // No key-value tuple exists for the specified key in the specified
+  // store.
+  no-such-key,
+
+  // Some implementation-specific error has occurred (e.g. I/O)
+  runtime(string)
+}
+
+// Open the store with the specified name.
+//
+// If `name` is the empty string, the default store is opened.  Otherwise,
+// `name` must refer to a store defined and configured in a runtime
+// configuration file supplied with the application.
+// `error::no-such-store` will be raised if the `name` is not recognized.
+open: func(name: string) -> expected<store, error>
+
+// Get the value associated with the specified `key` from the specified
+// `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store, and `error::no-such-key` will be raised if there is no
+// tuple for `key` in `store`.
+get: func(store: store, key: string) -> expected<list<u8>, error>
```

Comment:
Minor thing: could you please clarify what version of `.wit` is this document based on. I got confused for a second for the `exepcted` syntax as the current wit switches to `result` type but remembered the old wit syntax was using `expected`. 

Another comment is about the payload type from the `get` function. In wasi-keyvalue proposal, we realized that people might want differetn ways to consume the payload of a keyvalue store. This includes consuming it asynchronously and synchronously. Other that these, people might just want to examine the size of the payload, or some metadata of the payload. This observation leads to us proposing a payload `resource` type that includes some host functions for interpreting payload in different ways.

```fsharp
resource payload {
    consume_async: func() -> result<stream<u8>, Error>
    consume_sync: func() -> result<list<u8>, Error>
    size: func() -> result<u64, Error>
    // possibly more methods, like consume_async_with_key that returns a key payload pair `(key, stream<u8>)`
  }
```
Since `resource` is not there yet and this SIP is proposing using a `u32`-based resource, I would imagine it would difficult / non-ergonomic for users to call a different function to interpret the payload. But this is something I'd like your feedback on.

Reply:
Thanks for the feedback, @Mossaka.  I've added a note about the out-of-date WIT syntax (which we're unfortunately stuck on until we can update to the latest `wit-bindgen`) -- good catch.

I like the `payload` API you've proposed.  Perhaps we could have both:

```fsharp
// pseudo-handle for payloads:
type payload = u32
get: func(store: store, key: string) -> expected<list<u8>, error>
get-payload: func(store: store, key: string) -> expected<payload, error>
close-payload: func(payload: payload)
// pseudo-handle for streams:
type stream = u32
consume-async: func(payload: payload) -> expected<stream, error>
consume-sync: func(payload: payload) -> expected<list<u8>, error>
size: func(payload: payload) -> expected<u64, error>
// etc.
close-stream: func(stream: stream)
```
Then `get` would just be shorthand for calling `get-payload`, `consume-sync`, and finally `close-payload`.

Reply:
Given that this specific proposal is intentionally constrained to the wit-bindgen v0.2 feature set, is the payload pseudo-resource going to be worthwhile? We will never be able to implement `consume-async` in this version of the interface, for example.